sensor/common/scannerclient/grpc_client.go

```diff
@@ -2,72 +2,172 @@ package scannerclient
 	"github.com/stackrox/rox/pkg/logging"
 	"github.com/stackrox/rox/pkg/mtls"
 	"github.com/stackrox/rox/pkg/registries/types"
 	scannerV1 "github.com/stackrox/scanner/generated/scanner/api/v1"
 	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/status"
 )
 
 var (
 	log = logging.LoggerForModule()
 )
 
-// Client is a Scanner gRPC Client.
-type Client struct {
-	client scannerV1.ImageScanServiceClient
-	conn   *grpc.ClientConn
+// ScannerClient is the interface to remote image scanners used by Sensor.
+type ScannerClient interface {
+	GetImageAnalysis(context.Context, *storage.Image, *types.Config) (*ImageAnalysis, error)
+	Close() error
 }
 
-// dial Scanner and return a new Client.
-// dial is non-blocking and returns a non-nil error upon configuration error.
-func dial(endpoint string) (*Client, error) {
-	if endpoint == "" {
-		return nil, errors.New("Invalid Scanner endpoint (empty)")
+// ImageAnalysis is the result of an image analysis.
+type ImageAnalysis struct {
+	ScanStatus scannerV1.ScanStatus
+	ScanNotes  []scannerV1.Note
+	// Fields for analysis results for each supported Scanner: Scanner V2 (v1 proto)
+	// and Scanner V4 (v4 proto).
+	V1Components *scannerV1.Components
+	V4Contents   *v4.Contents
+}
+
+// v2Client is the client for StackRox Scanner based on Clair V2, also
+// known as Scanner V2.
+type v2Client struct {
+	scannerV1.ImageScanServiceClient
+	conn *grpc.ClientConn
+}
+
+// v4Client is the client for StackRox Scanner Indexer, also
+// known as Scanner V4 Indexer.
+type v4Client struct {
+	// Sensor uses Scanner V4's indexer.
+	v4.IndexerClient
+	conn *grpc.ClientConn
+}
+
+// GetStatus returns the image analysis status
+func (i *ImageAnalysis) GetStatus() scannerV1.ScanStatus {
+	if i != nil {
+		return i.ScanStatus
+	}
+	return scannerV1.ScanStatus_UNSET
+}
+
+// GetNotes returns the image analysis notes
+func (i *ImageAnalysis) GetNotes() []scannerV1.Note {
+	if i != nil {
+		return i.ScanNotes
+	}
+	return nil
+}
+
+// GetComponents returns the image analysis V1 components, available if the
+// underlying scan was done through legacy Scanner (aka. scanner-v2).
+func (i *ImageAnalysis) GetComponents() *scannerV1.Components {
+	if i != nil {
+		return i.V1Components
 	}
+	return nil
+}
+
+// GetContents returns the image analysis V4 contents, available if the
+// underlying scan was done through scanner-v4.
+func (i *ImageAnalysis) GetContents() *v4.Contents {
+	if i != nil {
+		return i.V4Contents
+	}
+	return nil
+}
 
-	endpoint = strings.TrimPrefix(endpoint, "https://")
-	if strings.Contains(endpoint, "://") {
-		return nil, errors.Errorf("Scanner endpoint has unsupported scheme: %s", endpoint)
+// getScannerEndpoint reads and validate the Scanner gRPC endpoint setting. If
+// the endpoint is empty or not configured properly (invalid) the value is
+// returned and error will be set.
+func getScannerEndpoint(s env.Setting) (string, error) {
```

Comment:
I think just passing in the `string` `s.Setting()` and making this function validate the endpoint (vs reading + validating) would be cleaner. I think we tend to pass in the string value of pre-evaluated env vars into functions instead of the `env.Setting`